### ISSUE TITLE
Issue when combining use_test_preprocessor and CMock's treat_inlines

### ISSUE THREAD

**Issue by i-adamov** [ThrowTheSwitch/Ceedling#706]
I am working on a big firmware project where the Ceedling is used to run unittests and mock driver headers. 
The issue I am having is that if I enable the test preprocessor feature and have CMock configured to treat header files with inline functions (of which we have some) the compilation of the unittest fails.

Additional info specific to our project is that the unittests (and project.yml file) are located in a separate directory so I am doing `CEEDLING_MAIN_PROJECT_FILE=./unittests/project.yml` before calling Ceedling.

I was able to recreate the issue using a simple example project that contains only a few files - https://github.com/i-adamov/ceedling-issue-example
I have a module which I need to test (`./src/example_file.c` and `./inc/example_file.h`) which includes a driver header (`./driverv/drv_bbb.h`) and in turn the driver header includes a HAL header (`./driver/hal/hal_aaa.h`). This simulates how our project is structured.
There is another header (`./inc/other_header.h`) which also includes the driver header and it is also included by the `example_file` module. It simulates the header of another module that may interact with the module I am testing.

When running Ceedling without the preprocessing it works fine:
`export CEEDLING_MAIN_PROJECT_FILE=./unittests/project.yml ; ceedling clobber test:all`
```
Clobbering all generated files...
(For large projects, this task may take a long time to complete)



Test 'test_unittest.c'
----------------------
Creating mock for drv_aaa...
Generating runner for test_unittest.c...
Compiling test_unittest_runner.c...
Compiling test_unittest.c...
Compiling mock_drv_aaa.c...
Compiling unity.c...
Compiling cmock.c...
Linking test_unittest.out...
Running test_unittest.out...


[==========] Running 2 tests from 1 test cases.
[----------] Global test environment set-up.   
[----------] 2 tests from test_unittest.c      
[ RUN      ] test_unittest.c.test_func1
[       OK ] test_unittest.c.test_func1 (0 ms)
[ RUN      ] test_unittest.c.test_static_func2
[       OK ] test_unittest.c.test_static_func2 (0 ms)
[----------] 2 tests from test_unittest.c (0 ms total)

[----------] Global test environment tear-down.
[==========] 2 tests from 0 test cases ran.
[  PASSED  ] 2 tests.
[  FAILED  ] 0 tests.

 0 FAILED TESTS

 0 FAILED TESTS
```

However if I set `:use_test_preprocessor: TRUE` the build fails:
```
Clobbering all generated files...
(For large projects, this task may take a long time to complete)



Test 'test_unittest.c'
----------------------
Generating include list for drv_aaa.h...
Creating mock for drv_aaa...
In file included from unittests/build/test/mocks/mock_drv_aaa.h:6:0,
                 from unittests/test/test_unittest.c:3:
unittests/build/test/mocks/drv_aaa.h:1:10: fatal error: driver/hal/hal_bbb.h: No such file or directory
 #include "driver/hal/hal_bbb.h"
          ^~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
ERROR: Shell command failed.
> Shell executed command:
'gcc -E -I"/var/lib/gems/2.5.0/gems/ceedling-0.31.1/vendor/unity/src" -I"/var/lib/gems/2.5.0/gems/ceedling-0.31.1/vendor/cmock/src" -I"unittests/build/test/mocks" -I"unittests/test" -I"unittests/test/support" -I"src" -I"inc" -I"driver/inc" -I"driver/hal" -D__STATIC_INLINE="static inline" -DTEST -D__STATIC_INLINE="static inline" -DTEST -DGNU_COMPILER "unittests/test/test_unittest.c" -o "unittests/build/test/preprocess/files/test_unittest.c"'
> And exited with status: [1].

rake aborted!
ShellExecutionException: ShellExecutionException
/var/lib/gems/2.5.0/gems/ceedling-0.31.1/lib/ceedling/tool_executor.rb:88:in `exec'
/var/lib/gems/2.5.0/gems/ceedling-0.31.1/lib/ceedling/preprocessinator_file_handler.rb:12:in `preprocess_file'
/var/lib/gems/2.5.0/gems/ceedling-0.31.1/lib/ceedling/preprocessinator.rb:48:in `preprocess_file'
/var/lib/gems/2.5.0/gems/ceedling-0.31.1/lib/ceedling/preprocessinator.rb:12:in `block in setup'
/var/lib/gems/2.5.0/gems/ceedling-0.31.1/lib/ceedling/preprocessinator_helper.rb:37:in `preprocess_test_file'
/var/lib/gems/2.5.0/gems/ceedling-0.31.1/lib/ceedling/preprocessinator.rb:33:in `preprocess_test_and_invoke_test_mocks'
/var/lib/gems/2.5.0/gems/ceedling-0.31.1/lib/ceedling/test_invoker.rb:84:in `block in setup_and_invoke'
/var/lib/gems/2.5.0/gems/ceedling-0.31.1/lib/ceedling/test_invoker.rb:51:in `setup_and_invoke'
/var/lib/gems/2.5.0/gems/ceedling-0.31.1/lib/ceedling/tasks_tests.rake:13:in `block (2 levels) in <top (required)>'
/var/lib/gems/2.5.0/gems/ceedling-0.31.1/bin/ceedling:345:in `block in <top (required)>'
/var/lib/gems/2.5.0/gems/ceedling-0.31.1/bin/ceedling:332:in `<top (required)>'
/usr/local/bin/ceedling:23:in `load'
/usr/local/bin/ceedling:23:in `<main>'
Tasks: TOP => test:all
(See full trace by running task with --trace)
ERROR: Ceedling Failed
```

What I see as a difference is that the driver header file (`unittests/build/test/mocks/drv_aaa.h`) which is processed by CMock to make the inline functions testable has changes to the include macros in the top of the file. It is also missing its include guard. 
When running without preprocessor: 
```
#ifndef DRV_AAA_H
#define DRV_AAA_H

#include "hal_bbb.h"

#define AAA 10

int get_aaa(void);

#endif
```
When running with preprocessor (empty lines truncated):
```
#include "driver/hal/hal_bbb.h"

int get_aaa(void);

```

The use of this `driver/hal/hal_bbb.h` filepath makes it impossible for the compiler to locate the file as the root directory is not used as an include path. However if I add it to the `:paths: :include:` section of the project.yml file, I get another issue with undefined macros:
```
Clobbering all generated files...
(For large projects, this task may take a long time to complete)



Test 'test_unittest.c'
----------------------
Generating include list for drv_aaa.h...
Creating mock for drv_aaa...
Generating runner for test_unittest.c...
Compiling test_unittest_runner.c...
Compiling test_unittest.c...
In file included from unittests/test/test_unittest.c:4:0:
src/example_file.c: In function 'func1':
inc/other_header.h:6:20: error: 'AAA' undeclared (first use in this function)
 #define SOMETHING (AAA + 5)
                    ^
src/example_file.c:13:24: note: in expansion of macro 'SOMETHING'
     return a + b + x + SOMETHING;
                        ^~~~~~~~~
inc/other_header.h:6:20: note: each undeclared identifier is reported only once for each function it appears in
 #define SOMETHING (AAA + 5)
                    ^
src/example_file.c:13:24: note: in expansion of macro 'SOMETHING'
     return a + b + x + SOMETHING;
                        ^~~~~~~~~
unittests/test/test_unittest.c: In function 'test_func1':
inc/other_header.h:6:20: error: 'AAA' undeclared (first use in this function)
 #define SOMETHING (AAA + 5)
                    ^
unittests/test/test_unittest.c:25:36: note: in expansion of macro 'SOMETHING'
     int expected = a + x * x + x + SOMETHING;
                                    ^~~~~~~~~
ERROR: Shell command failed.
> Shell executed command:
'gcc -I"/var/lib/gems/2.5.0/gems/ceedling-0.31.1/vendor/unity/src" -I"/var/lib/gems/2.5.0/gems/ceedling-0.31.1/vendor/cmock/src" -I"unittests/build/test/mocks" -I"unittests/test" -I"unittests/test/support" -I"src" -I"." -I"inc" -I"driver/inc" -I"driver/hal" -D__STATIC_INLINE="static inline" -DTEST -DGNU_COMPILER -g -c "unittests/test/test_unittest.c" -o "unittests/build/test/out/c/test_unittest.o" -MMD -MF "unittests/build/test/dependencies/test_unittest.d"'
> And exited with status: [1].

#<Thread:0x0000556d90ed87c8@/var/lib/gems/2.5.0/gems/ceedling-0.31.1/lib/ceedling/par_map.rb:7 run> terminated with exception (report_on_exception is true):
/var/lib/gems/2.5.0/gems/ceedling-0.31.1/lib/ceedling/tool_executor.rb:88:in `exec': ShellExecutionException (ShellExecutionException)
        from /var/lib/gems/2.5.0/gems/ceedling-0.31.1/lib/ceedling/generator.rb:99:in `generate_object_file'
        from /var/lib/gems/2.5.0/gems/ceedling-0.31.1/lib/ceedling/rules_tests.rake:17:in `block in <top (required)>'
        from /usr/lib/ruby/vendor_ruby/rake/task.rb:271:in `block in execute'
        from /usr/lib/ruby/vendor_ruby/rake/task.rb:271:in `each'
        from /usr/lib/ruby/vendor_ruby/rake/task.rb:271:in `execute'
        from /usr/lib/ruby/vendor_ruby/rake/task.rb:213:in `block in invoke_with_call_chain'
        from /usr/lib/ruby/2.5.0/monitor.rb:226:in `mon_synchronize'
        from /usr/lib/ruby/vendor_ruby/rake/task.rb:193:in `invoke_with_call_chain'
        from /usr/lib/ruby/vendor_ruby/rake/task.rb:182:in `invoke'
        from /var/lib/gems/2.5.0/gems/ceedling-0.31.1/lib/ceedling/task_invoker.rb:97:in `block in invoke_test_objects'
        from /var/lib/gems/2.5.0/gems/ceedling-0.31.1/lib/ceedling/par_map.rb:10:in `block (2 levels) in par_map'
rake aborted!
ShellExecutionException: ShellExecutionException
/var/lib/gems/2.5.0/gems/ceedling-0.31.1/lib/ceedling/tool_executor.rb:88:in `exec'
/var/lib/gems/2.5.0/gems/ceedling-0.31.1/lib/ceedling/generator.rb:99:in `generate_object_file'
/var/lib/gems/2.5.0/gems/ceedling-0.31.1/lib/ceedling/rules_tests.rake:17:in `block in <top (required)>'
/var/lib/gems/2.5.0/gems/ceedling-0.31.1/lib/ceedling/task_invoker.rb:97:in `block in invoke_test_objects'
/var/lib/gems/2.5.0/gems/ceedling-0.31.1/lib/ceedling/par_map.rb:10:in `block (2 levels) in par_map'
Tasks: TOP => unittests/build/test/out/c/test_unittest.o
(See full trace by running task with --trace)
ERROR: Ceedling Failed
```

What am I doing wrong? Do I need to enable some of the other Ceedling setting?
I need to be able to preprocess macros and to mock static functions in header files.

**Comment by M-Bab** [comment]
First of all thanks for writing this issue as it helped me to figure out my problem (which is exactly the same). I can confirm the issue and also provide info why this is the case and propose a solution.

Here is the problem step-by-step:
1. If `:treat_inlines: :include` is enabled, a modified header file needs to be generated (obviously because static/inline is removed from the header code as well). The modified header file can be found in `test/mocks` and is prioritized over the original header. The generation of this modified header happens in this line: https://github.com/ThrowTheSwitch/CMock/pull/261/files#diff-04520bc1e2e09e1dd300c2060876f6341f8b76093d60f845c52b8021bf36c5f1R66 as part of CMock.
2. If `:use_test_preprocessor: TRUE` is also enabled in Ceedling any header code provided to CMock is "filtered" through the actual GCC preprocessor. That preprocessor does a lot of stuff and among them is the removal of all macros - obviously because these macros are usually _applied_ by the GCC preprocessor.
3. As consequence of 1+2 you end up with a modified header file that is not only stripped of the keywords static/inline but also of all macros. When the GCC compiler finally tries to compile all the sources and any piece of code tries to use a macro of the original file it fails.

Here are two solutions that came to my mind:
1. Let the preprocessor do its job but prevent the removal of macros. The gcc preprocessor has extensive options (https://gcc.gnu.org/onlinedocs/gcc/Preprocessor-Options.html) where it might support anything like this, but I did not further dive into this. Also the ceedling option `use_preprocessor_directives` sounds pretty much like that, but did not resolve the problem described above.
2. Avoid feeding the preprocessor filtered file to the part of CMock which creates the modified header (without static/inline). Instead ensure this part gets the _original_ header file where it might strip out all static/inline keywords. Then the modified but not preprocessed file is stored under `test/mocks`. This is completely legitimate because once it actually compiles the stuff with gcc the preprocessor will run over the header again and is able to apply all the macros which are still stored. I will create a pull request where @mvandervoord can check if this is a good/okay solution.

@laurensmiers as the creator of the original feature (https://github.com/ThrowTheSwitch/CMock/pull/261) you might also be interested in this issue.

**Comment by Letme** [comment]
The 2. solution is not what you want.  The point of the use_test_preprocessor is to remove (well more like expand) the macros by preprocessor, so that CMock and other tools can correctly parse and generate only C valid (used) C code. In that part the 3. is actually what you would expect.

The top issue more points out that the path of the file included should be dependant on the preprocessed file and/or built file instead of original, or that inclusion of the file should actually be compiler argument include path dependant (not include directive dependant). It should also have include guard (which I agree is most probably the bug), but that will not solve the problem of removing the macro (which is what preprocessor does - not cmock, as you found out). So why do you want to use preprocessor if you do not want macros from header files to be removed?

**Comment by M-Bab** [comment]
> The 2. solution is not what you want. The point of the use_test_preprocessor is to remove (well more like expand) the macros by preprocessor, so that CMock and other tools can correctly parse and generate only C valid (used) C code. In that part the 3. is actually what you would expect.

I fear you didnt fully grasp my description. CMock **will** still get the correctly preprocessed header to operate normally. Only the very specific part of generating a **copy** of the original header without the static/inline keywords operates on the original file and not on the preprocessed file.

But I have to admit this issue is pretty much brainfuck.

**Comment by Letme** [comment]
So you do not forward the preprocessed file to the CMock for stripping static inline functions, but the original file is passed to the function?

**Comment by M-Bab** [comment]
My solution forwards both parts to CMock: The preprocessed header file _and_ the original file. All usual operations run as always in CMock only for the specific part of creating the copy with the static/inline keywords stripped the original file is fed.

Here are the decision tree options:
- :treat_inlines: :exclude (default) - No copy of the original header file is created. The original header file is used for compilation.
- :treat_inlines: :include and :use_test_preprocessor: FALSE - A copy of the original header with the static/include keywords stripped is created. This file works fine and still contains all the macros. It is obviously also what you get when you call CMock directly (without Ceedling there is no preprocessor option).
- :treat_inlines: :include and :use_test_preprocessor: TRUE - This is the error case. A copy of the original header is created where all the macros are removed. If the tests try to build and need any macros from the original header this build will fail.

**Comment by Letme** [comment]
And (check if I am correct):

 - :treat_inlines: :exclude (default) and use_test_preprocessor: TRUE - Creates preprocessed copy of a header file where all macros are removed

**Comment by M-Bab** [comment]
Yes and no. There is a copy created which is preprocessed but not what I consider as a copy in the description above. So not a copy which is placed **in the include path under `test/mocks`**.

Yes: There is a cache copy somewhere and the preprocessed data is passed to CMock.

No: This preprocessed header is not used when compiling the header later. When compiling the unittests the original header is used again and therefore naturally preprocessed by GCC and all macros are available. If this would be different the problem would be much bigger.

So to be more specific:

- :treat_inlines: :exclude (default) - No copy of the original header file is created **in the include path**. The original header file is used for compilation of the unittests. **This statement is valid independent of the use_test_preprocessor setting.**

**Comment by informatimago** [comment]
You may add -I. so that #include "driver/hal/hal_bbb.h" be successful.  You can do that in project.yml:
```
:paths:
  :source:
    - .
    - hal # etc
```
But it's strange and this shouldn't be needed.

**Comment by stemschmidt** [comment]
I am very surprised that treat_inlines creates a new header file even if the original header does not have any inline functions!?! 

I understand the need to create a header file if the inline implementation has to be mocked. But why is this a global setting and not depending on the content of the header?

**Comment by M-Bab** [comment]
Probably because its easier. Just do a batch processing of all files instead of actually looking into them and create a copy only if required.
But even the "smarter" behavior would still be fatal: Chances that important macros and inline functions are in the same header are pretty high.

**Comment by stemschmidt** [comment]
Adding to your solution 2 from January 12th 2023: Wouldn't it be good to have the strippables also be applied to the modified header? As far as I reverse engineered it this is only applied to the mocked files?

**Comment by M-Bab** [comment]
Phew ... the modified headers only exist for the sake of treat_inlines. I am not sure if my solution meddles with strippables in any way. I hope the behavior of strippables just stays exactly the same as without treat_inlines.

As also mentioned in my PRs: I am not very happy with the solution I created there. It was just the best shot I had in a few tries with very limited knowledge of ruby & rake.

**Comment by mkarlesky** [comment]
@i-adamov I _believe_ the latest prerelease of Ceedling 1.0.0 (formerly 0.32) fixes the problems documented in this issue. Ceedling's much improved preprocessing and CMock's `:treat_inlines` now work together as they should. I am going to leave this issue open to collect any followup. For anyone following this thread, please let us know if this problem has been corrected.

**Comment by M-Bab** [comment]
This is great news and will make Ceedling 1.0 the framework we are absolutely looking forward to. I will try this out as soon as I can (but this might need till begin of August) and provide feedback, if this works well in our setup.

**Comment by M-Bab** [comment]
Unfortunately the issue continues to exist according to the description with the latest ceedling version. Compare comment https://github.com/ThrowTheSwitch/Ceedling/issues/868#issuecomment-2397197627

**Comment by mkarlesky** [comment]
@i-adamov and @M-Bab So very sorry for the slow response. After staring at everything for a while and scratching my head, I _think_ I finally properly grasp the various problems at play. And, yes, this is not fixed as I thought it was. The current handling in 1.0.0 does work correctly for some limited cases but not the sophisticated cases typically at play here.

@M-Bab Your suggestion of using CMock to strip `inline`s and then perform preprocessing does certainly handle your case and many more cases than the current implementation. Unfortunately, it's not a universal solution. It's entirely plausible, and even likely in these sophisticated header files scenarios, where crazy macros are generating the problematic `inline` function signatures we want to handle specially. On the one hand, simple string manipulating like CMock uses will not handle all cases. On the other hand, running the preprocessor to get down to plain C code that we can work on blows away all macros with no clear path to getting them back (or preserving them in the first place).

If you have ideas, please share them. We're discussing how to handle all this before finally publishing 1.0.0.

I _think_ there _may_ be the appropriate options in the GCC preprocessor we could use with a scan-modify-stitch approach that gets us to 100% handling of all cases. But, I do not know yet.

**Comment by mkarlesky** [comment]
Referencing #938 by @MichaelBMiner to collect related issues here

**Comment by i-adamov** [comment]
Sorry for ghosting this tread for almost two years but I see @M-Bab has taken the torch and has done a great job of replicating the issue and tracking its progress. I try to keep up with the updates here but I've been too busy to try newer versions to see if the issue is fixed.
In my case, we just decided not to use the test preprocessor functionality we write our tests with this limitation in mind. We've had a bunch of problems that all come from this limitation but so far adding some modified or custom headers to the support dir in the unittests has been sufficient workaround.

**Comment by mkarlesky** [comment]
For all those playing along at home… @mvandervoord and I have looked at this problem long and hard now. We _**think**_ we have a solution that will handle all cases. We are working on implementing it and testing it now. If nothing else it should support everything but oddball edge cases.

This seems to be one of the most important advanced needs of power users, so, we think it's best to tackle this as the very last need before releasing 1.0.0. The whole idea of mocking is especially valuable to anyone working with complex header files, and `inline` usage is especially common in complex header files.

Wish us luck…

**Comment by M-Bab** [comment]
These were my MRs to cover the issue:
- https://github.com/ThrowTheSwitch/Ceedling/pull/728
- https://github.com/ThrowTheSwitch/CMock/pull/421

The trick there was to sneak the original header past the preprocessor to present it to the treat_inlines handling (and only to that feature).

**Comment by M-Bab** [comment]
> It's entirely plausible, and even likely in these sophisticated header files scenarios, where crazy macros are generating the problematic inline function signatures we want to handle specially.

Okay well yeah I see it now. If you use Macros to actually create your header lines you are in for a bad time. Not sure if this is a great style though. IDE parsers trying to help you with autocompletion might also crunch on that. So you are thinking of something like this:

```c
#define GENERATE_INLINE_FUNCTION_DECL(x) inline void (x)(void)

GENERATE_INLINE_FUNCTION_DECL(myfunc);
```

Can't remember if I have ever seen this. Most probably headers are rather auto generated with some scripts instead of doing such evil things. Well my suggestion here would probably be ...

```c
#if UNIT_TESTING
#define GENERATE_INLINE_FUNCTION_DECL(x) void (x)(void)
#else
#define GENERATE_INLINE_FUNCTION_DECL(x) inline void (x)(void)
#endif

GENERATE_INLINE_FUNCTION_DECL(myfunc);
```

But I am also curious about your generic solution which will handle it all. I will certainly test it with our Code base as a quite complex benchmark.

**Comment by mkarlesky** [comment]
@M-Bab Yup. Exactly. You got it on your example generator macro. I'll grant you that that's ugly enough that it causes a variety of other problems. If we absolutely possible that we may have to fall back to a less universal solution such as what you've proposed. Our unfortunate reality is that (A) embedded development is the primary use case for Ceedling (B) the embedded code Ceedling is asked to work with is often remarkably ugly with the craziest of provenance (C) the more universal any solution can be the more easily we can support other core needs in favor of time consuming edge cases (even just keeping up with questions).

So, we're gonna go for it with the solution we hope might work. If we're right, it'll handle even nutso cases — which means it will easily handle common cases like yours too. If it's not going to work we should know soon enough.

**Comment by mkarlesky** [comment]
Update: The first working version of the (hopefully) universal solution to this problem is in a branch and successfully running. It produces preprocessed header files that CMock is able to mock with optional handling of `inline` declarations. Any and all `#define` and `#pragma` statements in the source header file are preserved such that they remain in the copy of the header file CMock generates for `inline` handling. As a bonus, this fuller preprocessing technique is also able to preserve `TEST_SOURCE_FILE()` and `TEST_INCLUDE_PATH()` macros in test files (these were previously incompatible new features with Ceedling's preprocessing). Some sizable and long-needed memory optimizations are also in place.

This all needs some further cleanup, optimization, and documentation before it's ready to be fully exercised. I'm not sure when it will be available, but it looks like we're close to a proper solution.

**Comment by mkarlesky** [comment]
@M-Bab The [latest prerelease (1.0.0-13d0e3d)](https://github.com/ThrowTheSwitch/Ceedling/releases/tag/1.0.0-13d0e3d) is available for testing. It preserves macro definitions and pragmas in a header file that CMock copies and modifies when `:treat_inlines` is set to `:include`. This has all been fairly thoroughly unit testes and exercised with a project, but the real test is code from a real world, sophisticated project like what you have. Please give it a try at your earliest convenience and report on back. 🤞

**Comment by M-Bab** [comment]
Hi @mkarlesky ! Okay I did test it with `1.0.0-13d0e3d`. Unfortunately not successful yet. I can provide you a little guide-through without using too much of our internal code.

1. There is our rather simple osal.h header. Functions of it get used so it gets mocked in the test. [osal-mocked.zip](https://github.com/user-attachments/files/18041238/osal-mocked.zip)
2. Once it gets mocked with all the preprocessor functions active it creates a _new_ osal.h (together with mock_osal.h) in build/test/mocks/test_...
3. The amount of includes is dramatically reduced. Unfortunately it does not include `mdp_config.h` anymore. In there is a lot of configuration and among them the `OSAL_MUTEX` and `OSAL_SEMAPHORE` enums.
4. Building the test fails because the  types are unknown because `mdp_config.h` is not included anymore.

I hope this can somehow help you to do another step. I am certain that this was already one of the faults I got with the previous version I tested.

----

Slightly off-topic. I build Ceedling with a cloned repo and gem. But before ceedling actually worked I had to

```
gem install thor
gem install erb
```

Is it possibled that these modules are somehow missing in your requirements specifications?

**Comment by mkarlesky** [comment]
@M-Bab We did do some gem housekeeping recently. Thor is definitely in the Gemfile and gemspec. Not sure why it's failing. Erb is curious for other reasons. We will look into it.

**Comment by mkarlesky** [comment]
@M-Bab Could you kindly change the generated _osal.h_ to use the proper list of `#include` directives and then try to run your test or the (failing) compilation step within your build? I'd like to narrow in on what to fix. If the generated _osal.h_ includes the needed symbols it was previously lacking, then I know we only need to focus on the `#include` statements.

I can think of two options:

1. Temporarily alter your project `:cmock` configuration to inject all the missing headers and re-run one or more of your test files to completion.
1. Re-run your test file with verbosity elevated. Copy the compilation command line for your test file. Hand edit the generated (incomplete) _osal.h_ in the mock directory to reference the needed header files and manually compile your test at your terminal using the command line previously copied.

**Comment by M-Bab** [comment]
Hi again @mkarlesky ! I did the 2nd way - because I was not sure how to do the first option.

Good news here: The manual addition of `mdp_config.h` alone was sufficient to let the build step be successful. If you fix the cause that the includes are removed I will gladly test again. So far this is our simplest subproject we are unittesting but whenever this step is successful, I can get to more complex projects.

Merry christmas and a happy New Year.

**Comment by LuisAfonso95** [comment]
I've made some tests based on the issue I'd raised with #868  and it might be relevant here

I have to read out the process described above in more detail. However, I have made a test on my own with ST's hal library.

This is a fresh project with snapshot 1.0.0-13d0e3d. I have use_test_preprocessor: all and cmock with :treat_inlines:  :include  

This seems to work! I made a test using stm32h5xx_ll_gpio.h and stm32h5xx_ll_tim.h and it both works meaning

it mocks the inline functions in the headers when cmock_ is included in the test
mocks works as expected
the generated .h replacements have all the macros!
Given the state of the snapshot, this means I stop having TEST_CASE and TEST_RANGE from unity.
Tried use_test_preprocessor: mocks also works and gives me back the parametrization! I am excited to test this on the actual project at the start of the year :)

I want to test this one the big project first but for my specific use case this might be fixed.

Sorry for the delayed response, very busy few months.

**Comment by mvandervoord** [comment]
Thanks for the testing and feedback, @LuisAfonso95 . That's helpful!

@M-Bab -- I've just pushed a set of fixes that I believe will fix your remaining issue. Can you install the latest, then add `:use_deep_preprocessor: :mocks` to your project.yml file and give it another try when you have a moment?

```
:project:
  # optional features. If you don't need them, keep them turned off for performance
  :use_mocks: TRUE
  :use_test_preprocessor: :none  # options are :none, :mocks, :tests, or :all
  :use_deep_preprocessor: :none  # options are :none, :mocks, :tests, or :all
  :use_backtrace: :simple        # options are :none, :simple, or :gdb
  :use_decorators: :auto         # decorate Ceedling's output text. options are :auto, :all, or :none
```

**Comment by M-Bab** [comment]
Unfortunately no success so far :cry: 

I tested the following Ceedling versions just now:
- [test/ceedling_0_32_r](https://github.com/ThrowTheSwitch/Ceedling/tree/test/ceedling_0_32_rc)
- [1.0.0-88cea21](https://github.com/ThrowTheSwitch/Ceedling/releases/tag/1.0.0-88cea21)

I tested all the new options of `:use_deep_preprocessor` you made available `:none, :mocks, :tests and :all`. They all fail at the very same point. I am also confused here because the file itself is quite basic and simple. I recognized there are even more `osal.h` generated - not sure but it might be 2 more than before(?) - in the preprocess-subfolders `directives_only` and `full_expansion`. Both of these files still refer to the later missing `mdp_config.h`. These files are generated independent from the `:use_deep_preprocessor` setting.

Because it might help you, I created a new zip-file including all osal.h (the problematic file where the include statement is "lost") and our project.yml (don't get confused by out-commented code here this is still in the migration process for Ceedling 1.0.0):

[all-osal-h.zip](https://github.com/user-attachments/files/18274050/all-osal-h.zip)

I have absolutely no problem to continue testing for you. But I am not 100% sure if this is the fastest/most efficient approach. Maybe we can build a minimal problematic example project. Or we can hand over the relevant parts of our projects (would require an NDA I assume). Anyway both options would need some time. So I continue with the hope for the decisive breakthrough in this matter. :smiley:

**Comment by mkarlesky** [comment]
@M-Bab Well, darn. Thank you for sending the archive. We'll take a look.

In a previous exchange I mentioned adding the missing `#include` directives using your CMock config without explaining how to do that. Jump down to the `:includes:` section of the [CMock docs](https://github.com/ThrowTheSwitch/CMock/blob/master/docs/CMock_Summary.md). With a simple YAML list using any of the various `:include*` options, you can instruct CMock to inject additional `#include` statements in the mock files it generates.

```
:cmock:
  :includes: # With other variations you can control where in the order these additional files land
    - mdp_config.h
```

Just for sake of getting crystal clear on the problem, does adding the missing `#include` directives yourself this way get you to an otherwise successful test build?

**Comment by mkarlesky** [comment]
p.s. @M-Bab, On the gem install issue you mentioned previously, we think what you experienced is a few things coming together:

1. We did update the gem dependencies list recently, including updating minimum versions of existing dependencies.
1. Ruby itself has been pushing off certain previously core libraries as gems (e.g. `erb`).
1. Doing a local gem install (e.g. `gem install --local <ceedling .gem filepath>` or from a Gemspec in a cloned repo) does not actually process dependencies. Only installing from the RubyGems repository  does that (`gem install ceedling` — not yet available for 1.0.0). You could install the gem `bundler` and point its install process at the Gemspec in your cloned repo (`bundle install <local file options>`). Bundler _will_ install dependencies for you in a local installation that has access to the Gemspec.

So, it's likely that what happened is dependencies changed on you a bit and your local environment no longer matched the dependencies at the time of your local install. The normal way to fix it is to simply manually install the missing dependencies. This is expected. As I mentioned, Bundler is one way to handle this automatically.

I've updated the docs to address this (I did not get into the `bundler` business).

**Comment by M-Bab** [comment]
Yes, adding 

```
  :includes: # With other variations you can control where in the order these additional files land
    - mdp_config.h
```

to the cmocks section also makes this group of unit tests run successful. But it doesn't run the full controller successful yet. Most likely this needs a bunch of more specifically listed includes.

Sind I am switching quickly between Ceedling prerelease versions for these tests I built a little script :wink: 

```
#/!bin/bash
gem uninstall -x ceedling
rm -f ceedling-*.gem
gem build ceedling.gemspec
gem install ceedling-1.0.0.gem
ceedling help || exit 1
```

**Comment by mkarlesky** [comment]
@M-Bab Okay. Sounds like some progress!

Can you clarify a bit the distinction between “this group of unit tests” and “full controller”?

One theory I have is that there are conditional preprocessor statements sprinkled throughout the source code impacting the the chain of `#include` directives. It's _possible_ that everything is actually working, but the GCC preprocessor as Ceedling is calling it is unable to discover all the header files CMock needs to inject into its generated files. Why? Because one or more `#define`s are not set such that the preprocessor spiders to find the correct `#include` directives ultimately needed by Ceedlng and CMock to fully build up the mocks you need. Of course, this is just a theory. But, it wouldn't be the first time something like this has come up, and there's only grunt work to be done to find any missing `#define`s needed by a test build.

**Comment by M-Bab** [comment]
> Can you clarify a bit the distinction between “this group of unit tests” and “full controller”?

We unit test the complete product series which consists of multiple different embedded controllers which are tested. These tests are grouped by "modules" to test - where are module mostly matches a single source code file. With adding the include as suggested, I can complete the module of this controller successful. But still not all modules of this controller build successfully.

We barely use any precompile conditions on the include directives - at least in our code. And certainly not for this controller which has a quite simple code base.

But I don't have a great idea either where the problem is. But it worked quite well with my comparably simple solution from my 2 PRs. Is it maybe a fallback solution to support the two different modes (your sophisticated universal solution and my simple solution)? At least if we can't figure out the issues of the universal solution? This could avoid to delay the release due to the issue.

**Comment by LuisAfonso95** [comment]
hello, I mentioned before trying a 1.0 snapshot and inlines worked while keeping macros. Tried now the release with the actual full project and it doesn't keep macros anymore.

Quite weird, will have to compare with the small test project in my personal laptop

**Comment by LuisAfonso95** [comment]
As a follow up, I've checked my personal test. It indeed has all the defines when mocking with treat inlines and :use_test_preprocessor: :mocks. 

I even updated to release 1.0.0 instead of the snapshot and it works!
I will have to compare the project.yml. It must be some combo. I will check tomorrow

**Comment by mvandervoord** [comment]
That's good to hear. It'd be nice to understand why... but it sounds positive!  :)

**Comment by mkarlesky** [comment]
Hi, @LuisAfonso95. You got back to us while I was checking this. I just tried both a local clone of the repo and the Docker image we pushed out. It seems to be working for me as well. A copy of the mocked header file exists alongside the resulting mock with symbols and macros inserted at the top.

**Comment by LuisAfonso95** [comment]
> Hi, @LuisAfonso95. You got back to us while I was checking this. I just tried both a local clone of the repo and the Docker image we pushed out. It seems to be working for me as well. A copy of the mocked header file exists alongside the resulting mock with symbols and macros inserted at the top.

I've tried it again, and, TLDR, it worked. I checked the version and although it said Ceedling 1.0.0 instead of a snapshot I noticed Unity was 2.6.0 - it must have not installed correctly. Did and uninstall and install again and it worked in a small test project.

Tried it quickly on the real project and stm32h5xx_ll_gpio.h that was generated has all the macros!

**Comment by M-Bab** [comment]
@mvandervoord @mkarlesky Any updates on this issue? Anything more I can test for you?

**Comment by mkarlesky** [comment]
Sorry, @M-Bab. No further updates. We are trying to tackle a handful of pressing bugs and are spread a bit thin. Come to think of it, though… one of said bugs _**might**_ be the cause of what's failing for you. The 1.0.1 prerelease that is currently available has the fix for the problem I'm thinking of. If you'd be willing to try your project as-is with the latest 1.0.1 prerelease, that might tell us something more.

**Comment by M-Bab** [comment]
Yeah I can imagine there is a lot to do between the 1.0.0 and the 1.0.1 release. Just write me if I can be of any support.

I did another round of testing with 1.0.1-9283b04 and I can see a (slow) positive progress:

1. The first (simplest) controller can now build and run our tests. The setting `:use_deep_preprocessor: :mocks` seems completely irrelevant here - it works the same if the line is added or not. BTW is your documentation outdated here because ceedling itself stated only `mocks` and `none` are allowed settings now.
Key to success here was setting `:use_test_preprocessor: :tests`. It did not work with `:use_test_preprocessor: :all` but the error was not very reasonable (GCC error missing some declarations which are clearly in the first and only header that is included 😕 ).

2. I continued my testing with the next controller. The setting `:use_deep_preprocessor: :mocks` is also irrelevant here. The previously setting `:use_test_preprocessor: :tests` was not successful here. 

<details>

<summary>The internal parser gets stuck on a declaration from the NXP library: fsl_wdog32.h</summary>

```
🧨 EXCEPTION: CMock >> Failed Parsing Declaration Prototype!
  declaration: 'AT_QUICKACCESS_SECTION_CODE(void WDOG32_Init(WDOG_Type *base, const wdog32_config_t *config))'
  modifier: ''
  return: {:type => '', :name => 'cmock_to_return', :str => ' cmock_to_return', :void? => false, :ptr? => false, :const? => false, :const_ptr? => false}
  function: 'AT_QUICKACCESS_SECTION_CODE'
  args: [{:ptr? => false, :const? => false, :const_ptr? => false, :name => 'WDOG32_Init', :type => 'cmock_fsl_wdog32_func_ptr1'}]
🌱 Ceedling could not complete operations because of errors
```

</details>

<details>

<summary>The AT_QUICKACCESS_SECTION_CODE macro is declared as follows</summary>

```c
#if defined(FSL_SDK_DRIVER_QUICK_ACCESS_ENABLE) && FSL_SDK_DRIVER_QUICK_ACCESS_ENABLE
#if (defined(__ICCARM__))
#define AT_QUICKACCESS_SECTION_CODE(func) func @"CodeQuickAccess"
#define AT_QUICKACCESS_SECTION_DATA(func) func @"DataQuickAccess"
#elif(defined(__CC_ARM) || defined(__ARMCC_VERSION))
#define AT_QUICKACCESS_SECTION_CODE(func) __attribute__((section("CodeQuickAccess"))) func
#define AT_QUICKACCESS_SECTION_DATA(func) __attribute__((section("DataQuickAccess"))) func
#elif(defined(__GNUC__))
#define AT_QUICKACCESS_SECTION_CODE(func) __attribute__((section("CodeQuickAccess"))) func
#define AT_QUICKACCESS_SECTION_DATA(func) __attribute__((section("DataQuickAccess"))) func
#else
#error Toolchain not supported.
#endif /* defined(__ICCARM__) */
#else
#if (defined(__ICCARM__))
#define AT_QUICKACCESS_SECTION_CODE(func) func
#define AT_QUICKACCESS_SECTION_DATA(func) func
#elif(defined(__CC_ARM) || defined(__ARMCC_VERSION))
#define AT_QUICKACCESS_SECTION_CODE(func) func
#define AT_QUICKACCESS_SECTION_DATA(func) func
#elif(defined(__GNUC__))
#define AT_QUICKACCESS_SECTION_CODE(func) func
#define AT_QUICKACCESS_SECTION_DATA(func) func
#else
#error Toolchain not supported.
#endif
#endif /* __FSL_SDK_DRIVER_QUICK_ACCESS_ENABLE */
```

</details>

So for this controller I continued with the mode `:use_test_preprocessor: :all`. Until I was stopped by an *amazing* bug: Since the header-copies created under mocks for treat_inlines are preprocessed they might have a nuance of a difference to the original. Somehow they are both considered as includes, so you can end up with conflicting types/declarations between the files. More specifically this happened twice for me for structures which included arrays which size has a define instead of magic number. Which is obviously the better coding style.

<details>

<summary>Here an example extract of the original file</summary>

```c
#define MDP_ONE_WIRE_ADDR_SIZE 8u //!< Amount of bytes for a 1-wire device address

typedef struct
{
  MDP_GPIO pin;
  uint8_t  LastDiscrepancy;                /*!< Search private */
  uint8_t  LastFamilyDiscrepancy;          /*!< Search private */
  uint8_t  LastDeviceFlag;                 /*!< Search private */
  uint8_t  ROM_NO[MDP_ONE_WIRE_ADDR_SIZE]; /*!< 8-bytes address of last search device */
} MDP_ONE_WIRE;
```

</details>

<details>

<summary>This becomes transformed into this</summary>

```c
#define MDP_ONE_WIRE_ADDR_SIZE 8u

typedef struct
{
  MDP_GPIO pin;
  uint8_t LastDiscrepancy;
  uint8_t LastFamilyDiscrepancy;
  uint8_t LastDeviceFlag;
  uint8_t ROM_NO[8u];
} MDP_ONE_WIRE;
```

</details>

Although binary exactly the same this results in a large number of conflicting types and conflicting declarations.

**Comment by mkarlesky** [comment]
@M-Bab I still need to digest your message. Thank you for the update. This is all very helpful.

Quickly… the new `:use_deep_preprocessor` feature (we're still not sure if it's necessary) is intentionally limited to `:none` and `:mocks`. It has a subset of `:use_test_preprocessing` options for a complicated set of reasons.

**Comment by mkarlesky** [comment]
@M-Bab I've read over what you've sent along. Thank you.

First, I want to clarify something. I don't think I'm fully grasping some of your statements about `:use_test_preprocessor` settings. So, I'll just summarize the options to be certain:

* `:tests` causes preprocessing to only happen for your test files. Mockable headers are left as-is. With this setting, CMock's `inline` handling feature merely copies a header file and does simple text replacement on it (no actual toolchain preprocessing is involved).
* `:mocks` causes preprocessing to only happen for mockable headers. Test files are left as-is. With this setting, CMock's `inline` handling feature copies a header file and does simple text replacement on it after toolchain preprocessing and Ceedling processing.
* `:all` is both of the above. With respect to CMock's `inline` handling, the behavior is the same as `:mocks`.
* `:none` is self-explanatory.

If `:use_deep_preprocessing` is set to `:mocks` it comes into play if `:use_test_preprocessing` is `:mocks` or `:all`.

I can't say for certain, but it's possible the amazing bug you describe has been fixed in the latest prerelease dated January 26th. That's a bit of a long shot guess, though.

At this point, I would like to invite discussion on how to access your project within the bounds of your confidentiality requirements. Recent experiences with others on nuanced bugs since 1.0.0 has really driven home how invaluable access to code and build output is in finding problems. Days and weeks of back-and-forth with partial details, anonymized logs, and best-effort descriptions has often been reduced to just a half hour or less of direct investigation.

You've got a fantastic example of a real world project that stress tests some of Ceedling’s most sophisticated abilities. We're really not sure how to go forward from here. Even if we implemented the fallback option you've advocated, we have little confidence that this anything more than a bandaid instead of an actual fix. Our gut feeling is that Ceedling should be able to generically process your project, but it's obviously failing.

**Comment by mkarlesky** [comment]
We believe the problems at the heart of this issue have been addressed and are available in the 1.1.0 prerelease. I am going to close this issue. It can be reopened if needed.

One problem had to do with the ordering and assumptions of preprocessing to expand macros, etc. There were certain chicken-and-egg struggles. This problem has been fully resolved with the advent of Partials—see (2) below.

Another problem had to do with how preprocessing was discovering includes. In some cases the discovery went down a rabbit hole and extracted includes from OS, HAL, and system libraries. This was fixed by explicitly filtering out any includes discovered outside the collection of includes discovered from the project configuration.

In the meantime a couple important changes and fixes have also occurred:

1. Preprocessing has been refactored and greatly improved and optimized. The filtering discussed above is still present in this improved preprocessing.
1. [Ceedling Partials](https://throwtheswitch.github.io/Ceedling/latest/testing-guide/partials/) is a far richer option that can supersede CMock's more simplistic `treat_inlines` option. For many users Partials are probably a better and more robust route to achieve the same goal.
